tests/test_workers.py

```diff
@@ -0,0 +1,238 @@
+from __future__ import annotations
+
+import signal
+import subprocess
+import sys
+import tempfile
+import time
+from typing import TYPE_CHECKING
+
+import httpx
+import pytest
+
+if TYPE_CHECKING:
+    from ssl import SSLContext
+    from typing import IO, Generator
+
+    from asgiref.typing import (
+        ASGIReceiveCallable,
+        ASGISendCallable,
+        LifespanStartupFailedEvent,
+        Scope,
+    )
+
+pytestmark = pytest.mark.skipif(sys.platform == "win32", reason="requires unix")
+gunicorn_arbiter = pytest.importorskip("gunicorn.arbiter", reason="requires gunicorn")
+uvicorn_workers = pytest.importorskip("uvicorn.workers", reason="requires gunicorn")
+
+
+class Process(subprocess.Popen):
+    client: httpx.Client
+    output: IO[bytes]
+
+    def read_output(self) -> str:
+        self.output.seek(0)
+        return self.output.read().decode()
+
+
+@pytest.fixture(
+    params=(
+        pytest.param(uvicorn_workers.UvicornWorker, marks=pytest.mark.subprocess),
+        pytest.param(uvicorn_workers.UvicornH11Worker, marks=pytest.mark.subprocess),
+    )
+)
+def worker_class(request: pytest.FixtureRequest) -> str:
+    """Gunicorn worker class names to test.
+
+    This is a parametrized fixture. When the fixture is used in a test, the test
+    will be automatically parametrized, running once for each fixture parameter. All
+    tests using the fixture will be automatically marked with `pytest.mark.subprocess`.
+
+    https://docs.pytest.org/en/latest/how-to/fixtures.html
+    https://docs.pytest.org/en/latest/proposals/parametrize_with_fixtures.html
+    """
+    worker_class = request.param
+    return f"{worker_class.__module__}.{worker_class.__name__}"
+
+
+@pytest.fixture(
+    params=(
+        pytest.param(False, id="TLS off"),
+        pytest.param(True, id="TLS on"),
+    )
+)
+def gunicorn_process(
+    request: pytest.FixtureRequest,
+    tls_ca_certificate_pem_path: str,
+    tls_ca_ssl_context: SSLContext,
+    tls_certificate_private_key_path: str,
+    tls_certificate_server_cert_path: str,
+    unused_tcp_port: int,
+    worker_class: str,
+) -> Generator[Process, None, None]:
+    """Yield a subprocess running a Gunicorn arbiter with a Uvicorn worker.
+
+    An instance of `httpx.Client` is available on the `client` attribute.
+    Output is saved to a temporary file and accessed with `read_output()`.
+    """
+    app = "tests.test_main:app"
```

Comment:
We have several of these ASGI app instances around the test suite, so I just picked one. We could consider consolidating these app instances into one place in the test suite.

Reply:
I think it's better if we create the application on this file, following the same pattern we have everywhere in the test suite. Wdyt? 🤔

Reply:
Sure (1bc61dd).